modules/imgcodecs/src/loadsave.cpp

```diff
@@ -492,6 +532,136 @@ imread_( const String& filename, int flags, Mat& mat )
+    }else{
+#endif
+        decoder = findDecoder( filename, error );
+#ifdef HAVE_GDAL
+    }
+#endif
+
+    /// if no decoder was found, return nothing.
+    if( !decoder && error != IMREAD_OK){
+        return error;
+    }
+
+    int scale_denom = 1;
+    if( flags > IMREAD_LOAD_GDAL )
+    {
+        if( flags & IMREAD_REDUCED_GRAYSCALE_2 )
+            scale_denom = 2;
+        else if( flags & IMREAD_REDUCED_GRAYSCALE_4 )
+            scale_denom = 4;
+        else if( flags & IMREAD_REDUCED_GRAYSCALE_8 )
+            scale_denom = 8;
+    }
+
+    /// set the scale_denom in the driver
+    decoder->setScale( scale_denom );
+
+    /// set the filename in the driver
+    decoder->setSource( filename );
+
+    try
+    {
+        // read the header to make sure it succeeds
+        if( !decoder->readHeader() )
+            return IMREAD_ERROR_INVALID_HEADER;
+    }
+    catch (const cv::Exception& e)
+    {
+        std::cerr << "imread_('" << filename << "'): can't read header: " << e.what() << std::endl << std::flush;
+        return IMREAD_ERROR_INVALID_HEADER;
+    }
+    catch (...)
+    {
+        std::cerr << "imread_('" << filename << "'): can't read header: unknown exception" << std::endl << std::flush;
+        return IMREAD_ERROR_INVALID_HEADER;
+    }
+
+    // established the required input image size
+    Size size = validateInputImageSize(Size(decoder->width(), decoder->height()));
+
+    if(params.maxPixels != 0 && params.maxPixels <= static_cast<size_t>(decoder->width()) * decoder->height())
+    {
+        return IMREAD_ERROR_SIZE_LIMIT_EXCEEDED;
+    }
+
+    if(!params.maxSize.empty() && (params.maxSize.height < decoder->height() || params.maxSize.width < decoder->width()))
+    {
+        return IMREAD_ERROR_SIZE_LIMIT_EXCEEDED;
+    }
+
+    // grab the decoded type
+    int type = decoder->type();
+    if( (flags & IMREAD_LOAD_GDAL) != IMREAD_LOAD_GDAL && flags != IMREAD_UNCHANGED )
+    {
+        if( (flags & IMREAD_ANYDEPTH) == 0 )
+            type = CV_MAKETYPE(CV_8U, CV_MAT_CN(type));
+
+        if( (flags & IMREAD_COLOR) != 0 ||
+           ((flags & IMREAD_ANYCOLOR) != 0 && CV_MAT_CN(type) > 1) )
+            type = CV_MAKETYPE(CV_MAT_DEPTH(type), 3);
+        else
+            type = CV_MAKETYPE(CV_MAT_DEPTH(type), 1);
+    }
+
+    image.create( size.height, size.width, type );
+    Mat image_data = image.getMat();
+
+    // read the image data
+    bool success = false;
+    try
+    {
+        if (decoder->readData(image_data))
+            success = true;
+    }
+    catch (const cv::Exception& e)
+    {
+        std::cerr << "imread_('" << filename << "'): can't read data: " << e.what() << std::endl << std::flush;
+    }
+    catch (...)
+    {
+        std::cerr << "imread_('" << filename << "'): can't read data: unknown exception" << std::endl << std::flush;
+    }
+    if (!success)
+    {
+        image.release();
+        return IMREAD_ERROR_INVALID_DATA;
+    }
```

Comment:
I believe we should not "suppressing" of C++ exceptions. They provide much more useful information than single `ImreadError` value.
Searching for issue reports regarding to `IMREAD_ERROR_INVALID_DATA` keyword would be painful with many useless posts (as there is no useful information).

/cc @vpisarev